atomix/cluster/src/test/java/io/atomix/raft/RaftRule.java

```diff
@@ -0,0 +1,521 @@
+      }
+      servers.add(server);
+    }
+
+    latch.await(30 * nodes, TimeUnit.SECONDS);
+
+    return servers;
+  }
+
+  public void shutdownFollower() throws Exception {
+    final var follower = getFollower().orElseThrow();
+    shutdownServer(follower);
+  }
+
+  public RaftServer shutdownLeader() throws Exception {
+    final var leader = getLeader().orElseThrow();
+    shutdownServer(leader);
+    return leader;
+  }
+
+  public void restartLeader() throws Exception {
+    awaitNewLeader();
+    final var leader = shutdownLeader();
+
+    final RaftMember leaderMember = getRaftMember(leader.name());
+    createServer(leaderMember.memberId())
+        .join(getMemberIds())
+        .thenAccept(this::addCommitListener)
+        .get(30, TimeUnit.SECONDS);
+  }
+
+  private List<MemberId> getMemberIds() {
+    return members.stream().map(RaftMember::memberId).collect(Collectors.toList());
+  }
+
+  public void shutdownServer(final String memberId) throws Exception {
+    final var raftServer = getRaftServer(memberId);
+    shutdownServer(raftServer);
+  }
+
+  private RaftServer getRaftServer(final String memberId) {
+    return servers.stream()
+        .filter(server -> server.name().equals(memberId))
+        .findFirst()
+        .orElseThrow();
+  }
+
+  public void shutdownServer(final RaftServer raftServer) throws Exception {
+    raftServer.shutdown().get(30, TimeUnit.SECONDS);
+    servers.remove(raftServer);
+    memberLog.remove(raftServer.name());
+  }
+
+  public CompletableFuture<RaftServer> startServer(final String memberId) {
+    final RaftMember raftMember = getRaftMember(memberId);
+    final var server = createServer(raftMember.memberId());
+    return server.join(getMemberIds());
+  }
+
+  private RaftMember getRaftMember(final String memberId) {
+    return members.stream()
+        .filter(member -> member.memberId().id().equals(memberId))
+        .findFirst()
+        .orElseThrow();
+  }
+
+  public CompletableFuture<Void> tryToCompactLogsOnServersExcept(
+      final String memberId, final long index) {
+
+    final var servers =
+        this.servers.stream()
+            .filter(server -> !server.name().equals(memberId))
+            .collect(Collectors.toList());
+
+    final List<CompletableFuture<Void>> futures = new ArrayList<>();
+    for (final RaftServer server : servers) {
+      futures.add(tryToCompactLogOnServer(server, index));
+    }
+
+    return CompletableFuture.allOf(futures.toArray(new CompletableFuture[0]));
+  }
+
+  public CompletableFuture<Void> tryToCompactLogOnServer(final String memberId, final long index) {
+    final var raftServer = getRaftServer(memberId);
+    raftServer.getContext().getServiceManager().setCompactableIndex(index);
+    return raftServer.compact();
+  }
+
+  private CompletableFuture<Void> tryToCompactLogOnServer(
+      final RaftServer raftServer, final long index) {
+    raftServer.getContext().getServiceManager().setCompactableIndex(index);
```

Comment:
Is it safe to compact without a snapshot? Depending on what we want to test, this might not work.

Reply:
I think you're right. I have a follow up issue to investigate this.